util.go

```diff
@@ -47,16 +68,24 @@ type segmentedBuffer struct {
 }
 
 // NewSegmentedBuffer allocates a ring buffer.
-func NewSegmentedBuffer(initialCapacity uint32) segmentedBuffer {
+func newSegmentedBuffer(initialCapacity uint32) segmentedBuffer {
 	return segmentedBuffer{cap: initialCapacity, b: make([][]byte, 0)}
 }
 
+// Len is the amount of data in the receive buffer.
 func (s *segmentedBuffer) Len() int {
-	return int(atomic.LoadUint32(&s.len))
+	s.bm.Lock()
+	len := s.len
+	s.bm.Unlock()
+	return int(len)
 }
 
+// Cap is the remaining capacity in the receive buffer.
 func (s *segmentedBuffer) Cap() uint32 {
```

Comment:
we didn't do great on naming this, since `Cap` in go terminology gets used for the allocated space (len+cap) of the object. the comment helps though.

Reply:
Yeah. I tried making it work like go, but it wasn't much better.